src/order/modular_lattice.lean

```diff
@@ -1,41 +1,155 @@
 /-
 Copyright (c) 2020 Aaron Anderson. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Aaron Anderson
+Authors: Aaron Anderson, Yaël Dillies
 -/
-import order.rel_iso
+import order.cover
 import order.lattice_intervals
-import order.galois_connection
 
 /-!
 # Modular Lattices
-This file defines Modular Lattices, a kind of lattice useful in algebra.
+
+This file defines (semi)modular lattices, a kind of lattice useful in algebra.
 For examples, look to the subobject lattices of abelian groups, submodules, and ideals, or consider
 any distributive lattice.
 
+## Typeclasses
+
+We define (semi)modularity typeclasses as Prop-valued mixins.
+
+* `is_weak_up_modular_lattice`: Weakly upper modular lattices. Lattice where `a ⊔ b` covers `a` and
+  `b` if `a` and `b` both cover `a ⊓ b`.
+* `is_weak_low_modular_lattice`: Weakly lower modular lattices. Lattice where `a` and `b` cover
+  `a ⊓ b` if `a ⊔ b` covers both `a` and `b`
+* `is_up_modular_lattice`: Upper modular lattices. Lattices where `a ⊔ b` covers `a` if `b` covers
+  `a ⊓ b`.
+* `is_low_modular_lattice`: Lower modular lattices. Lattices where `a` covers `a ⊓ b` if `a ⊔ b`
+  covers `b`.
+- `is_modular_lattice`: Modular lattices. Lattices where `a ≤ c → (a ⊔ b) ⊓ c = a ⊔ (b ⊓ c)`. We
+  only require an inequality because the other direction holds in all lattices.
+
 ## Main Definitions
-- `is_modular_lattice` defines a modular lattice to be one such that
-  `x ≤ z → (x ⊔ y) ⊓ z ≤ x ⊔ (y ⊓ z)`
+
 - `inf_Icc_order_iso_Icc_sup` gives an order isomorphism between the intervals
   `[a ⊓ b, a]` and `[b, a ⊔ b]`.
   This corresponds to the diamond (or second) isomorphism theorems of algebra.
 
 ## Main Results
+
 - `is_modular_lattice_iff_inf_sup_inf_assoc`:
   Modularity is equivalent to the `inf_sup_inf_assoc`: `(x ⊓ z) ⊔ (y ⊓ z) = ((x ⊓ z) ⊔ y) ⊓ z`
 - `distrib_lattice.is_modular_lattice`: Distributive lattices are modular.
 
-## To do
-- Relate atoms and coatoms in modular lattices
+## References
+
+* [Wikipedia, *Modular Lattice*][https://en.wikipedia.org/wiki/Modular_lattice]
 
+## TODO
+
+- Relate atoms and coatoms in modular lattices
+- Prove that a modular lattice is both upper and lower modular.
 -/
 
 variable {α : Type*}
 
+/-- A weakly upper modular lattice is a lattice where `a ⊔ b` covers `a` and `b` if `a` and `b` both
+cover `a ⊓ b`. -/
+class is_weak_up_modular_lattice (α : Type*) [lattice α] : Prop :=
```

Comment:
Should use `upp` to indicate `upper` rather than `up` (which makes me expect the opposite to be `down` rather than `low`).
```suggestion
class is_weak_upp_modular_lattice (α : Type*) [lattice α] : Prop :=
```